sockeye/encoder.py

```diff
@@ -169,11 +216,44 @@ def encode(self, data: mx.sym.Symbol, data_length: mx.sym.Symbol, seq_len: int)
                                      input_dim=self.vocab_size,
                                      weight=self.embed_weight,
                                      output_dim=self.num_embed,
-                                     name=self.prefix + 'embed')
+                                     name=self.prefix + "embed")
+        if self.add_pos_encoding:
+            embedding = self._add_pos_encoding(embedding, seq_len)
         if self.dropout > 0:
             embedding = mx.sym.Dropout(data=embedding, p=self.dropout, name="source_embed_dropout")
         return embedding
 
+    def _add_pos_encoding(self, embedding: mx.sym.Symbol, seq_len: int) -> mx.sym.Symbol:
+        """
+        Adds positional encodings based on sine functions to embeddings.
+
+        Attention Is All You Need, Section 3.5
+        Vaswani et al. (https://arxiv.org/pdf/1706.03762.pdf)
+
+        :param embedding: Input embeddings (batch_size, seq_len, num_embed).
+        :param seq_len: Maximum sequence length.
+        :return: Summed input embeddings and position encodings (batch_size, seq_len, num_embed).
+        """
+
+        # (seq_len)
+        position = mx.sym.arange(start=0, stop=seq_len)
+        # (seq_len, 1)
+        position = mx.sym.expand_dims(data=position, axis=1)
+        pos_encoding_dims = []
+        for i in range(self.num_embed):
+            # (seq_len, 1)
+            pos_encoding_dim = mx.sym.sin(position / mx.sym.pow(10000, ((2 * i) / self.num_embed)))
+            pos_encoding_dims.append(pos_encoding_dim)
+        # (seq_len, num_embed)
+        pos_encoding = mx.sym.concat(*pos_encoding_dims, dim=1)
+        # (1, seq_len, num_embed)
+        pos_encoding = mx.sym.expand_dims(data=pos_encoding, axis=0)
+        # Position encodings are constant
+        pos_encoding = mx.sym.BlockGrad(pos_encoding)
```

Comment:
We should probably pre-compute the pos-encoding up to the maximum input length as to avoid recomputing the sin for each input. Then we can just slice and add.